### PR TITLE
Add support for DeepSeek language models in STORM Wiki pipeline

### PR DESCRIPTION
**Pull request by rmcc3**
## Description
This pull request adds support for using DeepSeek language models in the STORM Wiki pipeline, providing an alternative to the existing OpenAI models. The integration allows users to easily switch between OpenAI and DeepSeek models, enhancing the flexibility and capabilities of the STORM Wiki system.

## Key Changes

1. Implemented a new `DeepSeekModel` class in `src/lm.py`, which is compatible with the existing `dspy.OpenAI` interface.
2. Created a new example script `examples/run_storm_wiki_deepseek.py` to demonstrate how to use DeepSeek models with the STORM Wiki pipeline.
3. Updated the `STORMWikiRunner` class to be model-agnostic, ensuring compatibility with both OpenAI and DeepSeek models.
4. Added topic name sanitization to handle special characters in file names, which fixes a bug which would cause the research to crash when using special characters.
5. Add Unicode normalization to handle non-ASCII characters.

## How to Use

Users can now run the STORM Wiki pipeline with DeepSeek models by using the new `run_storm_wiki_deepseek.py` script. The script allows configuration of various DeepSeek-specific parameters such as model choice, temperature, and top_p sampling. The API key should be set with `DEEPSEEK_API_KEY`. The DeepSeek base can be set with `DEEPSEEK_API_BASE`, if needed.

## Testing
* Tested the integration with DeepSeek models.
* Verified compatibility with existing retrieval methods.
* Ensured proper handling of topic names with special characters.

## Future Considerations
* Consider implementing a unified interface for easily switching between different model providers and retrievers.

### CODE REVIEW

File: examples/run_storm_wiki_deepseek.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        raise ValueError("DEEPSEEK_API_KEY environment variable is not set. Please set it in your secrets.toml file.")
+
+    deepseek_kwargs = {
+        'api_key': os.getenv("DEEPSEEK_API_KEY"),
+        'api_base': os.getenv("DEEPSEEK_API_BASE", "https://api.deepseek.com"),
+        'temperature': args.temperature,
+        'top_p': args.top_p,
+    }
+
+    # DeepSeek offers two main models: 'deepseek-chat' for general tasks and 'deepseek-coder' for coding tasks
```
Comment (shaoyijia) [inline examples/run_storm_wiki_deepseek.py:65]: Hi @rmcc3 , thanks for the contribution! It looks pretty good.

Could you provide an input-output example (i.e., the generated article) using `deepseek-chat` or `deepseek-coder`? In this way, we will be more confident to merge this PR. Thank you so much!

Comment (rmcc3) [reply examples/run_storm_wiki_deepseek.py:65]: Sure thing!

Google was ran with Deepseek chat, and Stackoverflow was ran with Deepseek coder.
[Google.zip](https://github.com/user-attachments/files/16287525/Google.zip)
[Stackoverflow.zip](https://github.com/user-attachments/files/16287528/Stackoverflow.zip)

google
Stackoverflow

File: src/lm.py [hunk 2]
```diff
@@ -107,6 +107,94 @@ def __call__(
...
+        assert return_sorted is False, "for now"
+
+        response = self._create_completion(prompt, **kwargs)
+
+        # Log the token usage from the DeepSeek API response.
+        self.log_usage(response)
+
+        choices = response["choices"]
+        completions = [choice["message"]["content"] for choice in choices]
+
```
Comment (shaoyijia) [inline src/lm.py:196]: Add the following lines to make sure the call history is tracked:

```python
history = {
    "prompt": prompt,
    "response": response,
    "kwargs": kwargs,
}
self.history.append(history)
```

**Review by shaoyijia — CHANGES_REQUESTED** [review]
Hi @rmcc3 , thank you so much for the prompt response! The quality of the example outputs is pretty good - we are very happy to support DeepSeek models in our project. Could you make the following change so that I can merge this PR?
- Resolve the merge conflict. (We recently have breaking change in order to support installing our project via `pip`.)
- Add a few lines (see my comment) to `DeepSeekModel` to ensure the call history is tracked. Currently, llm_call_history.jsonl is empty in your shared output. After adding these lines, it shall include call history in the session.

**Comment by rmcc3** [comment]
Seems updating to head caused this to close. Working on the change now.

**Comment by rmcc3** [comment]
[llm_call_history.txt](https://github.com/user-attachments/files/16308847/llm_call_history.txt)

Here is the call history (had to upload as `.txt` because GitHub does not accept `.jsonl`)

**Pull request merged**